File: 1. Basic Topics/2. Backtracking/cpp/78. Subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> tmp;
        sort(nums.begin(), nums.end());
        backtrack(ans, tmp, nums,  0);
        return ans;
    }

    void backtrack(vector<vector<int>>& ans, vector<int>& tmp, vector<int>& nums, int start) {
        ans.push_back(tmp);    
        for(int i = start; i < nums.size(); i++) {
            tmp.push_back(nums[i]);
            backtrack(ans, tmp, nums, i + 1);
            tmp.pop_back();
        }
    }
};

File: 1. Basic Topics/2. Backtracking/py/78. Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        tmp = []
        nums.sort()
        self.backtrack(ans, tmp, nums, 0)
        return ans
    
    def backtrack(self, ans: List[List[int]], tmp: List[int], nums: List[int], start: int) -> None:
        ans.append(list(tmp))  
        for i in range(start, len(nums)):
            tmp.append(nums[i])
            self.backtrack(ans, tmp, nums, i + 1)
            tmp.pop() 